docs/dsl/example.md

```diff
@@ -0,0 +1,438 @@
+# DSL example
+
+Audience: Datalab developers.
+
+Status: rough/ready, somewhat aspirational, uses old cohort terminology.  I would like to explore whether this can be used as the basis for some of the DSL documentation.
+
+---
+
+Imagine we're studying covid vaccinations, and we're interested in patients who have been registered at the same practice for a full year.
+
+We want to end up with a one-row-per-patient table which we can then use for our analysis (an "input.csv" table), like the following:
+
+| patient | age | sex | first_date | first_code | second_date | second_code | codes_match |
+| - | - | - | - | - | - | - | - |
+| 1234 | 42 | f | 2021-03-19 | pf | 2021-07-04 | pf | yes |
+| 2345 | 51 | m | - | - | - | - | - |
+| 4567 | 61 | m | 2021-06-23 | az | - | - |
+| 5678 | 29 | f | 2021-07-01 | pf | 2021-12-27 | az | no |
+
+The data we need lives in three database tables:
+
+**patients**
+
+| patient | age | sex |
+| - | - | - |
+| 1234 | 42 | f |
+| 2345 | 51 | m |
+| 3456 | 83 | f |
+| 4567 | 61 | m |
+| 5678 | 29 | f |
+
+**registrations**
+
+| id | patient | practice | start_date | end_date |
+| - | - | - | - | - |
+| 78 | 1234 | A | 1989-03-12 | 2021-04-13 |
+| 79 | 1234 | B | 2012-04-13 | - |
+| 80 | 2345 | C | 1999-12-01 | - |
+| 81 | 3456 | D | 2001-12-03 | 2017-09-09 |
+| 82 | 4567 | E | 2019-06-01 | 2015-09-01 |
+| 83 | 4567 | F | 2015-10-25 | - |
+| 84 | 5678 | G | 2019-02-12 | - |
+
+
+**immunisations**
+
+| id | patient | code | date |
+| - | - | - | - |
+| 43 | 1234 | pf | 2021-03-19 |
+| 44 | 4567 | pox | 2021-10-10 |
+| 45 | 4567 | az | 2021-06-23 |
+| 46 | 1234 | flu | 2021-10-15 |
+| 47 | 1234 | pf | 2021-07-04 |
+| 48 | 2345 | flu | 2021-03-25 |
+| 49 | 2345 | flu | 2021-08-19 |
+| 50 | 5678 | pox | 2021-10-11 |
+| 51 | 4567 | flu | 2021-12-14 |
+| 52 | 3456 | pf | 2021-09-09 |
+| 53 | 2345 | flu | 2021-11-02 |
+| 54 | 5678 | pf | 2021-07-01 |
+| 55 | 3456 | pox | 2021-11-03 |
+| 56 | 3456 | az | 2021-04-01 |
+| 57 | 5678 | az | 2021-12-30 |
+| 58 | 5678 | pf | 2021-07-02 |
+
+**patients** is a `PatientFrame`, while **registrations** and **immunisations** are `EventFrame`s.
+
+---
+
+A `PatientFrame` represents a collection of rows with at most one row per patient.  An `EventFrame` represents as collection of rows with multiple rows per patient.
+
+An `EventFrame`/`PatientFrame` can be filtered to produce another `EventFrame`/`PatientFrame`, and you should think of them as being like unevaluated Django Querysets, in that they can be combined and acted on to produce something that can be turned into SQL.
+
+Ultimately, we transform an `EventFrame`/`PatientFrame` into a `PatientSeries`, through filtering, sorting, aggregating, and selecting a column.
```

Comment:
suggestion: I was half-asleep and didn't realise the pandas inspiration until @inglesp mentioned it when talking through the DSL.

It's possibly useful to note that: the naming and data representation is inspired by analogy to pandas because of the data representation. But these are specific data structures here and aren't actually pandas `pd.Series` or `pd.DataFrame`.